src/main/java/htsjdk/samtools/cram/structure/SubstitutionMatrix.java

```diff
@@ -164,44 +140,125 @@ public SubstitutionMatrix(final byte[] matrix) {
         baseByCode['N'][(encodedMatrixBytes[4] >> 2) & 3] = 'G';
         baseByCode['N'][(encodedMatrixBytes[4]) & 3] = 'T';
 
-        for (final byte refBase : SUBSTITUTION_BASES_UPPER) {
+        for (final SubstitutionBase refBase : BASES) {
             for (byte code = 0; code < CODES_PER_BASE; code++) {
-                codeByBase[refBase][baseByCode[refBase][code]] = code;
+                codeByBase[refBase.getBase()][baseByCode[refBase.getBase()][code]] = code;
             }
         }
     }
 
+    /**
+     * Given a reference base and a read base, find the corresponding substitution code
+     * @param refBase reference base being substituted
+     * @param readBase read base to substitute for the reference base
+     * @return code to be used for this refBase/readBase pair
+     */
+    public byte code(final byte refBase, final byte readBase) {
+        if (refBase <= 0 || Character.isLowerCase((char) refBase)) {
```

Comment:
What if the refBase is either upper-case non valid base e.g. 'J' or some symbol chracter like '<'... How is that we care so much so censore lower-case letters (not only valid bases) but nothing else that in fact is not a valid base call?

Since we go to so much trouble to create this big 128, 128 matrices we could use them to detect bad ref bases without calling isLowerCase.... if (codeByBase[refBase] == null is a bad refBase... that cause a throw... and only then we can check whether refBase was a lower-case version of an otherwise valid base in which case you customize the error message accordingly. 

Alternatively, if we accepted '>' or 'J' as "good" or "ok" ref bases what would happen to these case down-the-road... was an exception thrown on them later one, perhaps we should use the same approach with 'a', 'c', 'g', 't' and  'n'.

Reply:
You're correct that the parameter checking here is somewhat selective. Previously this code would have silently returned a code given any pair of byte values, and I've added only selective checks that reject completely nonsensical inputs (such as `refBase <= 0`), or attempts to use lower case bases (since we're unclear why code was specifically included to handle them when they're not allowed by the spec, and want to find out if we have any code paths or data that depends on it).

More thoroughly checking of the reference and read bases *is* something we should do, but it should be done at the site(s) where `Substitution` class instances are created in the first place. The CRAM refactoring is being done in a series PRs rather than all at once in order to keep the reviews manageable and prevent them from getting stale. The `Substitution` changes, will be in one of the next refactoring rounds, along with changes to use more strongly typed arguments for these methods, i.e., `Substitution` instead of (`byte, byte`), after which the matrix indexing will be completely safe.

Reply:
For you are saying here you actually added this additional lower-case check but at the same time is not clear what we should be doing about them in the end (after those several PRs)... or simply the problem will disappear by using specialized input types rather than standard byte primitives ....  in that case wouldn't make more sense simply error on the side of simplicity and omit this change i.e. don't check for lower-case as it used to be?

Reply:
@vruano Initially I was only logging a warning here, since I still want a have a way to find out if there are cases where we are using lower case bases (especially, other places in the code that aren't normalizing the bases to upper case, as they should be). Changing it so it throws was based on review feedback, and I think it makes sense. Its the surest way to find any offending code paths. So I think throwing for now is the right thing to do.